blackhole enable 20 cores

bjanjicTT opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/26056

### Problem description
5x4 blackhole grid is required for porting and testing OFT, Panoptic and SSR models as an emulation of BOS chip. 
Also enable tests to be executed under blackhole-post-commit.yaml  

### What's changed
Added an env variable to override **compute_with_storage_grid_range: end:.** param from yaml file from https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/core_descriptors/
example of usage: CORE_GRID_OVERRIDE_TODEPRECATE="4,3"
Added a fast division operation with a divisor of 20
New yaml file for a 5x4 grid tests under a blackhole-post-comit.yaml test execution.

### Checklist
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)

bjanjicTT committed:
blackhole enable 20 cores

bjanjicTT committed:
adding spdf copyright license header

williamlyTT commented on .github/workflows/blackhole-post-commit.yaml:
```diff
@@ -330,6 +330,15 @@ jobs:
+    needs: build-artifact
+    secrets: inherit
+    uses: ./.github/workflows/blackhole-20-cores-tests.yaml
+    with:
+      runner-label: ${{ inputs.runner-label || 'BH' }}
```
Please use the job matrix like in other single card workflows

```
    needs: [build-artifact, generate-matrix]
    secrets: inherit
    uses: ./.github/workflows/blackhole-20-cores-tests.yaml
    strategy:
      fail-fast: false
      matrix:
        test-group: ${{ fromJson(needs.generate-matrix.outputs.civ2-matrix) }}
    with:
      arch: blackhole
      runner-label: ${{ matrix.test-group }}
```

williamlyTT commented on .github/workflows/blackhole-20-cores-tests.yaml:
```diff
@@ -0,0 +1,73 @@
+      - name: Run ${{ matrix.test-group.name }}
+        env:
+          ARCH_NAME: ${{ matrix.test-group.arch }}
+          BH_ARCH_YAML: blackhole_20_arch.yaml
```
@tenstorrent/metalium-developers-infra what are our thoughts on `BH_ARCH_YAML` since we just removed `WH_ARCH_YAML`?

afuller-TT replied:
Kill it with fire.

What is the reasoning for it?  WH_ARCH_YAML was painful and took way too long to remove.  It finally happened.  Why repeat that history?

williamlyTT reviewed (dismissed):
Please make the runner label changes and run blackhole post-commit

tt-rkim commented on tt_metal/llrt/core_descriptor.cpp:
```diff
@@ -70,6 +70,15 @@ inline std::string get_core_descriptor_file(
                                             ? "wormhole_b0_80_arch_eth_dispatch.yaml"
                                             : "wormhole_b0_80_arch.yaml");
             case tt::ARCH::BLACKHOLE:
+                if (getenv("BH_ARCH_YAML")) {
```
I wonder if we need a larger conversation on how to let consumers specify their own core descriptor. or at least choose from an array of options we have. Then they could programmatically choose via the API.

abhullar-tt replied:
yes @tt-rkim yamls like this shouldn't be needed, rather move towards specified config approach https://github.com/tenstorrent/tt-metal/issues/26474

tt-rkim replied:
Why can't we then expand the conversation to see what's possible and prioritize the cleanup work?

We just did a huge cleanup effort that took almost 2 years to remove `WH_ARCH_YAML`. We do NOT want another instance of that.

tt-rkim replied:
Those who suffer the most are those who must maintain the testing of it.

mbezuljTT replied:
All valid concerns and I share them with you. 

BOS codebase is months/weeks behind ours, and to move quickly we really need this kind of feature. 

We also need to enable op owners to easily test their ops (existing and new ones) on the specific core grid. 

@abhullar-tt how much of work do you think specified config approach would be? 

If it's weeks - I can take technical debt on my team to move every test/model we create while it's done. 

Please share your thoughts!

tt-rkim replied:
We'll follow up on https://github.com/tenstorrent/tt-metal/issues/26474
later as teams to ensure this is being worked on

Just want the environment variable renamed, thanks

blozano-tt reviewed (dismissed):
Please explore all other alternatives before adding yet another environment variable.

We aim to end environment variable engineering here at Tenstorrent.

It took 1 year to remove WH_ARCH_YAML and it just got removed 4 days ago.

https://github.com/tenstorrent/tt-metal/issues/11059

cc: @handrewsTT

bjanjicTT committed:
fupdated env variable and matrix generation

bjanjicTT committed:
added changes for runner

bjanjicTT committed:
arch

bjanjicTT committed:
"P150b workflow run fix"

bjanjicTT commented:
Hi @tt-rkim  I have fixed env variable name as agreed in https://github.com/tenstorrent/tt-metal/issues/26474#issuecomment-3180592347 and addressed comment from @williamlyTT. 
My team also agreed that we will clean depth once Almeet's team provide solution for this. You can see post-commit here https://github.com/tenstorrent/tt-metal/actions/runs/16938905413 and once it is done could you please dismiss Wilders review and merge this PR?

bjanjicTT committed:
Update .github/workflows/blackhole-20-cores-tests.yaml

Co-authored-by: William Ly <williamly@tenstorrent.com>

bjanjicTT committed:
Merge branch 'main' into bjanjic/enable-20-cores-on-blackhole

bjanjicTT committed:
Merge branch 'main' into bjanjic/enable-20-cores-on-blackhole

TT-BrianLiu commented on tests/ttnn/unit_tests/test_bh_20_cores_sharding.py:
```diff
@@ -0,0 +1,59 @@
+from tests.ttnn.utils_for_testing import assert_with_pcc
+
+
+def skip_if_not_blackhole_20_cores(device):
+    if not is_blackhole():
+        pytest.skip("This test is intended to run only on Blackhole devices with 20 cores.")
+    compute_grid = device.compute_with_storage_grid_size()
+    if compute_grid.x != 5 or compute_grid.y != 4:
+        pytest.skip(
+            f"This test is intended to run only on Blackhole devices with 20 cores. Core grid [{compute_grid.x},{compute_grid.y}] must be [5, 4]."
+        )
+
+
+def test_sharding_on_bh_20_cores(device):
+    skip_if_not_blackhole_20_cores(device)
```
idk, until something is captured concretely in code, might be better to just parse the env var: `BH_ARCH_YAML_TODEPRECATE`

bjanjicTT replied:
As long as the grid is 4x5 and blackhole is used, we expect the test to pass.

bjanjicTT committed:
Merge branch 'main' into bjanjic/enable-20-cores-on-blackhole

abhullar-tt commented:
> In my mind, env variable that can set any grid size (smaller than a full grid) exposes unlimited options for use and has a high risk to increase technical debt much further than 3 models than we are doing for BOS and that are reason we raised this change.
> 
> @tt-rkim from infra perspective what is your preference?

I see what you mean but the reason I proposed this was because we need a similar way to allow users to fallback to 8x8 core grid on Galaxy now and thought the same env var could be used (rather than a second temp env var). We could even constrain it to only allow the temp env var to select 8x8 for galaxy and 5x4 on BH. The real fix for all of this would be to uplift Ops to natively handle SubDevice and uplift our MetalContext initialization but those are a couple weeks out

mbezuljTT commented:
I understand your point of view. 

To me it appears as gluing two hacks together, but I don't really mind - I never vouched for the removal of WH_ARCH_YAML. 

Please approve PR or return for edits, so we are clear with the next steps.

abhullar-tt commented:
> I understand your point of view.
> 
> To me it appears as gluing two hacks together, but I don't really mind - I never vouched for the removal of WH_ARCH_YAML.
> 
> Please approve PR or return for edits, so we are clear with the next steps.

I am suggesting to get rid of the new core desc yaml and add an env var to override the parsed core grid and make it 5x4

bjanjicTT committed:
removed blackhole-20-cores.yaml and introduced CORE_GRID_OVERRIDE_TODEPRECATE env variable

bjanjicTT committed:
coment updated

bjanjicTT committed:
comment updates

bjanjicTT committed:
removed file from cmake

bjanjicTT committed:
missing yaml params

abhullar-tt reviewed (approved):
sorry for all the back and forth, thanks for the changes!

bjanjicTT committed:
env var moved to rtoptions

abhullar-tt reviewed (approved):
thanks for making all the changes

bjanjicTT committed:
Merge branch 'main' into bjanjic/enable-20-cores-on-blackhole

bjanjicTT committed:
Merge branch 'main' into bjanjic/enable-20-cores-on-blackhole

bjanjicTT committed:
Merge branch 'main' into bjanjic/enable-20-cores-on-blackhole

bjanjicTT committed:
Merge branch 'main' into bjanjic/enable-20-cores-on-blackhole
